satpy/readers/msi_safe.py

```diff
@@ -194,14 +214,37 @@ def band_indices(self):
     @cached_property
     def band_offsets(self):
         """Get the band offsets from the metadata."""
-        offsets = self.root.find(".//Radiometric_Offset_List") if self.process_level == "L1C" else \
+        offsets = self.root.find(".//Radiometric_Offset_List") if self.process_level[:2] == "L1" else \
             self.root.find(".//BOA_ADD_OFFSET_VALUES_LIST")
         if offsets is not None:
             band_offsets = {int(off.attrib["band_id"]): float(off.text) for off in offsets}
         else:
             band_offsets = {}
         return band_offsets
 
+    def solar_irradiance(self, band_name):
+        """Get the solar irradiance for a given *band_name*."""
+        band_index = self._band_index(band_name)
+        return self.solar_irradiances[band_index]
+
+    @cached_property
+    def solar_irradiances(self):
+        """Get the TOA solar irradiance values from the metadata."""
+        irrads = self.root.find(".//Solar_Irradiance_List")
+        if irrads is not None:
+            solar_irrad = {int(irr.attrib["bandId"]): float(irr.text) for irr in irrads}
+        else:
+            solar_irrad = {}
+        return solar_irrad
+
+    @cached_property
+    def sun_earth_dist(self):
+        """Get the sun-earth distance from the metadata."""
+        sed = self.root.find(".//U")
+        if sed is not None:
+            return float(sed.text)
+        return -1
```

Comment:
I'm not really found of this. Would it be possible to raise an error here and catch it downstream?